### ISSUE TITLE
[doc] Produce rule documentation with properties

### ISSUE THREAD

**Issue by adangel** [pmd/pmd#1712]
Part of #1139

Right now, we produce a small code snippet, that one can copy+paste into their own ruleset to use a rule, e.g. on [ClassNamingConventions](https://pmd.github.io/pmd-6.12.0/pmd_rules_java_codestyle.html#classnamingconventions) we provide this snippet:

```xml
<rule ref="category/java/codestyle.xml/ClassNamingConventions" />
```

Maybe we can extend this to also provide here the properties with their default values, e.g.

```xml
<rule ref="category/java/codestyle.xml/ClassNamingConventions">
    <properties>
        <property name="classPattern" value="[A-Z][a-zA-Z0-9]*" />
        <property name="abstractClassPattern" value="[A-Z][a-zA-Z0-9]*" />
        <property name="interfacePattern" value="[A-Z][a-zA-Z0-9]*" />
        <property name="enumPattern" value="[A-Z][a-zA-Z0-9]*" />
        <property name="annotationPattern" value="[A-Z][a-zA-Z0-9]*" />
        <property name="utilityClassPattern" value="[A-Z][a-zA-Z0-9]+(Utils?\|Helper)" />
    </properties>
</rule>
```

**Pro**:
* That way, it's more obvious, that and how you can customize the rule.
* If someone copies this snippet including the default values, they won't be affected, if we change the default. (note: this is also con...)

**Con**:
* If someone copies this snippet including the default values, they won't be affected, if we change the default. (note: this is also pro...)

**Neutral**:
* The users still need to read the documentation. If they don't know, how to customize the rulesets, they would still run into the problems such as #1710.
* We could even go that far (for PMD 7 at least), to require also property value override. The default would then be merely a documentation suggestion. This would (maybe?) force users to think about what they are configuring? But still, if the snippet is just copied, then it's the same as having implicit default values.

**Comment by jsotuyod** [comment]
I wouldn't change the behavior of PMD 7, as forcing boilerplate is pretty bad for most users. If we have sane defaults, customizations should be minimal.

I do however love adding such blocks alongside the existing configuration snippet.

This is something Checkstyle for instance is currently doing (and wasn't not so long ago):

http://checkstyle.sourceforge.net/config_annotation.html#SuppressWarnings_Examples

**Comment by oowekyala** [comment]
I agree with @jsotuyod on this. Checkstyle example configs are very nice, I think we should do the same. Maybe adding an `example-config` element to the schema for 7.0.0, and generating a default one (the current one, with just the reference and no properties).

**Comment by jsotuyod** [comment]
@oowekyala do we really need this in the schema? I think we can autogenerate it in the docs just as we do the properties' table, without extra work on the rulesets.

**Comment by oowekyala** [comment]
I was thinking, that adding an element would allow us to describe non-trivial configurations for specific use cases instead of just showing a config with all properties (with defaults) and a config with no properties. That way we can be more specific, giving examples with non-defaults, and describing how the config changes the behaviour of the rule. Something like:
```xml
<example-config>
  <description>This configuration would ignore methods tagged with the "@org.foo.Whatever" annotation. </description>
  <rule-ref ref="category/java/SomeRule" >
    <properties>
       <property name="ignoredAnnotations">
          <seq>
            <item>org.foo.Whatever</item>
          </seq>
       </property>
    </properties>
  </rule-ref>
</example-config>
```

Or for naming conventions rules, give other common naming conventions that wouldn't be appropriate as default values (eg the `myX` convention for fields).

**Comment by jsotuyod** [comment]
Ohh, I see… on a deeper look it may make sense… Checkstyle seems to work somewhat like that, even including code snippets for each configuration

http://checkstyle.sourceforge.net/config_annotation.html#AnnotationLocation_Examples

**Comment by adangel** [comment]
Yes, sounds useful for 7.0.0. Since it's changing the schema, it's definitively for 7.0.0 only.
I've looked into checkstyle, and their documentation seems not to be generated, e.g. https://github.com/checkstyle/checkstyle/blob/55971426f06594ac5b7a529832ce639bc09d5722/src/xdocs/config_annotation.xml#L135

What about simply generating an additional section with the default values in PMD 6?
So, e.g. for https://pmd.github.io/pmd-6.12.0/pmd_rules_java_codestyle.html#classnamingconventions
we would generate the documentation like this:

---

## ClassNamingConventions

**Since:** PMD 1.2

**Priority:** High (1)

Configurable naming conventions for type declarations. This rule reports
type declarations which do not match the regex that applies to their
specific kind (e.g. enum or interface). Each regex can be configured through
properties.

By default this rule uses the standard Java naming convention (Pascal case),
and reports utility class names not ending with 'Util'.

**This rule is defined by the following Java class:** [net.sourceforge.pmd.lang.java.rule.codestyle.ClassNamingConventionsRule](https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/ClassNamingConventionsRule.java)

**Example(s):**

``` java
// This is Pascal case, the recommended naming convention in Java
// Note that the default values of this rule don't allow underscores 
// or accented characters in type names
public class FooBar {}

// You may want abstract classes to be named 'AbstractXXX',
// in which case you can customize the regex for abstract
// classes to 'Abstract[A-Z]\w+'
public abstract class Thing {}

// This class doesn't respect the convention, and will be flagged
public class Éléphant {}
```

**This rule has the following properties:**

|Name|Default Value|Description|Multivalued|
|----|-------------|-----------|-----------|
|classPattern|\[A-Z\]\[a-zA-Z0-9\]\*|Regex which applies to concrete class names|no|
|abstractClassPattern|\[A-Z\]\[a-zA-Z0-9\]\*|Regex which applies to abstract class names|no|
|interfacePattern|\[A-Z\]\[a-zA-Z0-9\]\*|Regex which applies to interface names|no|
|enumPattern|\[A-Z\]\[a-zA-Z0-9\]\*|Regex which applies to enum names|no|
|annotationPattern|\[A-Z\]\[a-zA-Z0-9\]\*|Regex which applies to annotation names|no|
|utilityClassPattern|\[A-Z\]\[a-zA-Z0-9\]+(Utils?\|Helper)|Regex which applies to utility class names|no|

**Use this rule with the default properties by just referencing it:**
``` xml
<rule ref="category/java/codestyle.xml/ClassNamingConventions" />
```

**Use this rule and customize it:**
```xml
<rule ref="category/java/codestyle.xml/ClassNamingConventions">
    <properties>
        <property name="classPattern" value="[A-Z][a-zA-Z0-9]*" />
        <property name="abstractClassPattern" value="[A-Z][a-zA-Z0-9]*" />
        <property name="interfacePattern" value="[A-Z][a-zA-Z0-9]*" />
        <property name="enumPattern" value="[A-Z][a-zA-Z0-9]*" />
        <property name="annotationPattern" value="[A-Z][a-zA-Z0-9]*" />
        <property name="utilityClassPattern" value="[A-Z][a-zA-Z0-9]+(Utils?\|Helper)" />
    </properties>
</rule>
```

---

**Comment by jsotuyod** [comment]
@adangel I'd go for it in 6.x, and see how people start reacting before introducing larger changes.

**Comment by oowekyala** [comment]
Isn't this closed? @jsotuyod you reopened it immediately, maybe this is a misclick?

**Comment by adangel** [comment]
Closed by #1714 for PMD 6.13.0.
